src/mlpack/methods/regularized_svd/regularized_svd_function_impl.hpp

```diff
@@ -186,5 +218,109 @@ double StandardSGD::Optimize(
   return overallObjective;
 }
 
+
+template <>
+template <>
+inline double ParallelSGD<ExponentialBackoff>::Optimize(
+    mlpack::svd::RegularizedSVDFunction<arma::mat>& function,
+    arma::mat& iterate)
+{
+  double overallObjective = DBL_MAX;
+  double lastObjective;
+
+  // The order in which the functions will be visited.
+  arma::Col<size_t> visitationOrder = arma::linspace<arma::Col<size_t>>(0,
+      (function.NumFunctions() - 1), function.NumFunctions());
+
+  const arma::mat data = function.Dataset();
+
+  // Iterate till the objective is within tolerance or the maximum number of
+  // allowed iterations is reached. If maxIterations is 0, this will iterate
+  // till convergence.
+  for (size_t i = 1; i != maxIterations; ++i)
+  {
+    // Calculate the overall objective.
+    lastObjective = overallObjective;
+    overallObjective = 0;
+
+    #pragma omp parallel for reduction(+:overallObjective)
+    for (omp_size_t j = 0; j < (omp_size_t) function.NumFunctions(); ++j)
+    {
+      overallObjective += function.Evaluate(iterate, j);
+    }
+
+    // Output current objective function.
+    Log::Info << "Parallel SGD: iteration " << i << ", objective "
+      << overallObjective << "." << std::endl;
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Parallel SGD: converged to " << overallObjective
+        << "; terminating with failure. Try a smaller step size?"
+        << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SGD: minimized within tolerance " << tolerance << "; "
+        << "terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    // Get the stepsize for this iteration
+    double stepSize = decayPolicy.StepSize(i);
+
+    // Shuffle for uniform sampling of functions by each thread.
+    std::random_shuffle(visitationOrder.begin(), visitationOrder.end());
```

Comment:
hm, the difference is still minimal, we are talking about nanoseconds. Just did some experiments using:

```
void ShuffelArma(arma::Col<size_t>& data)
{
  auto start = chrono::steady_clock::now();

  data = std::move(arma::shuffle(data));

  auto end = chrono::steady_clock::now();
  auto diff = end - start;

  typedef std::chrono::duration<float> floatSeconds;
  auto secs = std::chrono::duration_cast<floatSeconds>(diff);

  cout << "arma shuffel: " << chrono::duration <double, milli> (diff).count() << " ms - " << secs.count() << " sec" << endl;
}

void ShuffelStd(arma::Col<size_t>& data)
{
  auto start = chrono::steady_clock::now();

  std::random_shuffle(data.begin(), data.end());

  auto end = chrono::steady_clock::now();
  auto diff = end - start;

  typedef std::chrono::duration<float> floatSeconds;
  auto secs = std::chrono::duration_cast<floatSeconds>(diff);

  cout << "std shuffel: " << chrono::duration <double, milli> (diff).count() << " ms - " << secs.count() << " sec" << endl;
}
```

my results:

```
size: 10
std shuffel: 0.012911 ms - 1.2911e-05 sec
arma shuffel: 0.018432 ms - 1.8432e-05 sec
---------------------------------
size: 100
std shuffel: 0.006969 ms - 6.969e-06 sec
arma shuffel: 0.008546 ms - 8.546e-06 sec
---------------------------------
size: 1000
std shuffel: 0.063325 ms - 6.3325e-05 sec
arma shuffel: 0.071951 ms - 7.1951e-05 sec
---------------------------------
size: 10000
std shuffel: 0.664872 ms - 0.000664872 sec
arma shuffel: 0.751739 ms - 0.000751739 sec
---------------------------------
size: 100000
std shuffel: 6.71557 ms - 0.00671557 sec
arma shuffel: 8.78779 ms - 0.00878779 sec
---------------------------------
size: 1000000
std shuffel: 66.6233 ms - 0.0666233 sec
arma shuffel: 92.9679 ms - 0.0929679 sec
---------------------------------
size: 10000000
std shuffel: 891.552 ms - 0.891552 sec
arma shuffel: 1077.56 ms - 1.07756 sec
---------------------------------
size: 100000000
std shuffel: 10738.9 ms - 10.7389 sec
arma shuffel: 12356.7 ms - 12.3567 sec
---------------------------------
```

So for 100000000 elements the difference is about 1.6 seconds, which I think is still minimal, but I don't mind to use ``std::random_shuffle`` instead of ``arma::shuffel``.